main.go

```diff
@@ -1684,8 +1684,24 @@ func main() {
 			usage(command)
 			os.Exit(1)
 		}
-		if options.Target == "" && filepath.Ext(outpath) == ".wasm" {
-			options.Target = "wasm"
+		if options.Target == "" {
+			switch {
+			case options.GOARCH == "wasm":
+				switch options.GOOS {
+				case "js":
+					options.Target = "wasm"
+				case "wasip1":
+					options.Target = "wasip1"
+				case "wasip2":
+					options.Target = "wasip2"
+				default:
+					fmt.Fprintln(os.Stderr, "GOARCH=wasm but GOOS is unset. Please set GOOS to wasm, wasip1, or wasip2.")
+					os.Exit(1)
```

Comment:
There's also the case of `GOARCH=wasm GOOS=foobar`, which would produce a "GOOS is unset" error. That's somewhat wrong (though hopefully not very common). To be really sure it might be better to have a `""` case.